File: src/types/science-events.ts
/**
 * TypeScript Type Definitions for Science Event Store
 *
 * These types match the backend Python schema and enable type-safe
 * integration with the Science Event Store API.
 */

// ============================================================================
// Core Event Types
// ============================================================================

export type EventType =
  | "CLINICAL_READOUT"
  | "MECHANISM_INSIGHT"
  | "EVIDENCE_UPDATE"
  | "REGULATORY_CHANGE"
  | "TRIAL_UPDATE"
  | "TARGET_VALIDATION"
  | "GENETIC_EVIDENCE"
  | "BIOMARKER_DISCOVERY"
  | "PARTNERSHIP"
  | "ACQUISITION"
  | "COMPETITIVE_INTEL"
  | "SAFETY_SIGNAL"
  | "ENDPOINT_CHANGE"
  | "APPROVAL"
  | "REJECTION"
  | "LABEL_UPDATE";

export type EventCategory =
  | "CLINICAL"
  | "PRECLINICAL"
  | "REGULATORY"
  | "MECHANISM"
  | "COMMERCIAL";

export type EvidenceClass =
  | "GENETIC"
  | "PRECLINICAL"
  | "TRANSLATIONAL"
  | "CLINICAL"
  | "RWE"
  | "REGULATORY";

export type EntityType =
  | "DRUG"
  | "COMPANY"
  | "TARGET"
  | "INDICATION"
  | "TRIAL";

export type SourceType =
  | "FDA"
  | "CT.gov"
  | "EMA"
  | "SEC"
  | "PUBMED"
  | "CLINICALTRIALS"
  | "PRESS_RELEASE"
  | "CONFERENCE"
  | "UNKNOWN";

export type RelationshipType =
  | "FOLLOWS"
  | "PRECEDES"
  | "CAUSES"
  | "RESULTS_FROM"
  | "CONTRADICTS"
  | "SUPPORTS"
  | "REFINES"
  | "UPDATES"
  | "INVALIDATES";

// ============================================================================
// Science Event Schema
// ============================================================================

export interface RelatedEntity {
  type: EntityType;
  id: string;
  name: string;
}

export interface KeyFinding {
  finding: string;
  significance?: string;
  [key: string]: any;
}

export interface ScienceEvent {
  id: number;
  event_type: EventType;
  event_category?: EventCategory;
  title: string;
  description?: string;
  summary?: string;

  // Temporal
  event_date: string;  // ISO datetime
  published_date?: string;  // ISO date

  // Entity associations
  entity_type?: EntityType;
  entity_id?: string;
  entity_name?: string;
  related_entities?: RelatedEntity[];

  // Source and provenance
  source_type?: SourceType;
  source_url?: string;
  source_metadata?: Record<string, any>;

  // Content
  content?: string;
  key_findings?: KeyFinding[];
  impact_assessment?: string;

  // Classification
  evidence_class?: EvidenceClass;
  confidence_score?: number;  // 0-1
  impact_score?: number;  // 0-1

  // Versioning
  version: number;
  parent_version_id?: number;
  is_current: boolean;
  change_summary?: string;

  // Metadata
  tags?: string[];
  event_metadata?: Record<string, any>;

  // Timestamps
  created_at: string;  // ISO datetime
  updated_at?: string;  // ISO datetime
}

// ============================================================================
// Event Relationship Schema
// ============================================================================

export interface EventRelationship {
  id: number;
  source_event_id: number;
  target_event_id: number;
  relationship_type: RelationshipType;
  description?: string;
  confidence?: number;  // 0-1
  event_metadata?: Record<string, any>;
  created_at: string;  // ISO datetime
}

// ============================================================================
// API Request Types
// ============================================================================

export interface CreateScienceEventRequest {
  event_type: EventType;
  event_category?: EventCategory;
  title: string;
  description?: string;
  summary?: string;
  event_date: string;
  published_date?: string;
  entity_type?: EntityType;
  entity_id?: string;
  entity_name?: string;
  related_entities?: RelatedEntity[];
  source_type?: SourceType;
  source_url?: string;
  source_metadata?: Record<string, any>;
  content?: string;
  key_findings?: KeyFinding[];
  impact_assessment?: string;
  evidence_class?: EvidenceClass;
  confidence_score?: number;
  impact_score?: number;
  tags?: string[];
  metadata?: Record<string, any>;
}

export interface UpdateScienceEventRequest extends CreateScienceEventRequest {
  change_summary?: string;
}

export interface CreateEventRelationshipRequest {
  source_event_id: number;
  target_event_id: number;
  relationship_type: RelationshipType;
  description?: string;
  confidence?: number;
  metadata?: Record<string, any>;
}

export interface ListEventsQuery {
  event_type?: EventType;
  event_category?: EventCategory;
  entity_type?: EntityType;
  entity_id?: string;
  source_type?: SourceType;
  evidence_class?: EvidenceClass;
  from_date?: string;  // ISO datetime
  to_date?: string;  // ISO datetime
  tags?: string;  // Comma-separated
  min_confidence?: number;
  min_impact?: number;
  current_only?: boolean;
  limit?: number;
  offset?: number;
}

export interface TimelineQuery {
  from_date?: string;
  to_date?: string;
  event_types?: string;  // Comma-separated
}

// ============================================================================
// API Response Types
// ============================================================================

export interface CreateEventResponse {
  id: number;
  event_type: EventType;
  title: string;
  event_date: string;
  created_at: string;
}

export interface ListEventsResponse {
  events: ScienceEvent[];
  total: number;
  limit: number;
  offset: number;
}

export interface TimelineResponse {
  entity_type: EntityType;
  entity_id: string;
  timeline: ScienceEvent[];
  total_events: number;
}

export interface EventHistoryVersion {
  id: number;
  version: number;
  title: string;
  event_date: string;
  is_current: boolean;
  change_summary?: string;
  created_at: string;
}

export interface EventHistoryResponse {
  versions: EventHistoryVersion[];
}

export interface EventRelationshipsResponse {
  event_id: number;
  relationships: EventRelationship[];
}

export interface AggregationByType {
  event_type: EventType;
  count: number;
}

export interface AggregateResponse {
  aggregations: AggregationByType[];
}

export interface SearchResult {
  id: number;
  event_type: EventType;
  title: string;
  summary?: string;
  event_date: string;
  entity_type?: EntityType;
  entity_id?: string;
  source_type?: SourceType;
}

export interface SearchResponse {
  query: string;
  results: SearchResult[];
  count: number;
}

// ============================================================================
// Client API
// ============================================================================

export class ScienceEventStoreAPI {
  constructor(private baseUrl: string = '/api/v1/science') {}

  /**
   * Create a new science event
   */
  async createEvent(event: CreateScienceEventRequest): Promise<CreateEventResponse> {
    const response = await fetch(`${this.baseUrl}/science-events`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(event)
    });

    if (!response.ok) {
      throw new Error(`Failed to create event: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * List science events with filtering
   */
  async listEvents(query?: ListEventsQuery): Promise<ListEventsResponse> {
    const params = new URLSearchParams();
    if (query) {
      Object.entries(query).forEach(([key, value]) => {
        if (value !== undefined) {
          params.append(key, String(value));
        }
      });
    }

    const response = await fetch(`${this.baseUrl}/science-events?${params}`);

    if (!response.ok) {
      throw new Error(`Failed to list events: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Get a specific science event
   */
  async getEvent(eventId: number): Promise<ScienceEvent> {
    const response = await fetch(`${this.baseUrl}/science-events/${eventId}`);

    if (!response.ok) {
      throw new Error(`Failed to get event: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Update a science event (creates new version)
   */
  async updateEvent(
    eventId: number,
    event: UpdateScienceEventRequest
  ): Promise<CreateEventResponse> {
    const response = await fetch(`${this.baseUrl}/science-events/${eventId}`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(event)
    });

    if (!response.ok) {
      throw new Error(`Failed to update event: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Get version history of an event
   */
  async getEventHistory(eventId: number): Promise<EventHistoryResponse> {
    const response = await fetch(`${this.baseUrl}/science-events/${eventId}/history`);

    if (!response.ok) {
      throw new Error(`Failed to get event history: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Get timeline of events for an entity
   */
  async getTimeline(
    entityType: EntityType,
    entityId: string,
    query?: TimelineQuery
  ): Promise<TimelineResponse> {
    const params = new URLSearchParams();
    if (query) {
      Object.entries(query).forEach(([key, value]) => {
        if (value !== undefined) {
          params.append(key, String(value));
        }
      });
    }

    const response = await fetch(
      `${this.baseUrl}/science-events/timeline/${entityType}/${entityId}?${params}`
    );

    if (!response.ok) {
      throw new Error(`Failed to get timeline: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Create a relationship between events
   */
  async createRelationship(
    relationship: CreateEventRelationshipRequest
  ): Promise<{ id: number }> {
    const response = await fetch(`${this.baseUrl}/event-relationships`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(relationship)
    });

    if (!response.ok) {
      throw new Error(`Failed to create relationship: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Get relationships for an event
   */
  async getRelationships(
    eventId: number,
    relationshipType?: RelationshipType,
    direction: 'incoming' | 'outgoing' | 'both' = 'both'
  ): Promise<EventRelationshipsResponse> {
    const params = new URLSearchParams({ direction });
    if (relationshipType) {
      params.append('relationship_type', relationshipType);
    }

    const response = await fetch(
      `${this.baseUrl}/event-relationships/${eventId}?${params}`
    );

    if (!response.ok) {
      throw new Error(`Failed to get relationships: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Aggregate events by type
   */
  async aggregateByType(
    fromDate?: string,
    toDate?: string,
    entityType?: EntityType
  ): Promise<AggregateResponse> {
    const params = new URLSearchParams();
    if (fromDate) params.append('from_date', fromDate);
    if (toDate) params.append('to_date', toDate);
    if (entityType) params.append('entity_type', entityType);

    const response = await fetch(
      `${this.baseUrl}/science-events/aggregate/by-type?${params}`
    );

    if (!response.ok) {
      throw new Error(`Failed to aggregate events: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Search events by text
   */
  async searchEvents(query: string, limit: number = 50): Promise<SearchResponse> {
    const params = new URLSearchParams({ q: query, limit: String(limit) });

    const response = await fetch(`${this.baseUrl}/science-events/search?${params}`);

    if (!response.ok) {
      throw new Error(`Failed to search events: ${response.statusText}`);
    }

    return response.json();
  }
}

// ============================================================================
// React Hooks (if using React)
// ============================================================================

import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query';

export function useScienceEvents(query?: ListEventsQuery) {
  const api = new ScienceEventStoreAPI();

  return useQuery({
    queryKey: ['science-events', query],
    queryFn: () => api.listEvents(query)
  });
}

export function useScienceEvent(eventId: number) {
  const api = new ScienceEventStoreAPI();

  return useQuery({
    queryKey: ['science-event', eventId],
    queryFn: () => api.getEvent(eventId),
    enabled: !!eventId
  });
}

export function useEventTimeline(entityType: EntityType, entityId: string, query?: TimelineQuery) {
  const api = new ScienceEventStoreAPI();

  return useQuery({
    queryKey: ['timeline', entityType, entityId, query],
    queryFn: () => api.getTimeline(entityType, entityId, query)
  });
}

export function useCreateEvent() {
  const api = new ScienceEventStoreAPI();
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: (event: CreateScienceEventRequest) => api.createEvent(event),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['science-events'] });
    }
  });
}

export function useUpdateEvent(eventId: number) {
  const api = new ScienceEventStoreAPI();
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: (event: UpdateScienceEventRequest) => api.updateEvent(eventId, event),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['science-event', eventId] });
      queryClient.invalidateQueries({ queryKey: ['science-events'] });
    }
  });
}

export function useEventSearch(query: string, limit?: number) {
  const api = new ScienceEventStoreAPI();

  return useQuery({
    queryKey: ['event-search', query, limit],
    queryFn: () => api.searchEvents(query, limit),
    enabled: query.length >= 3
  });
}

// ============================================================================
// Helper Functions
// ============================================================================

/**
 * Format event date for display
 */
export function formatEventDate(dateString: string): string {
  return new Date(dateString).toLocaleDateString('en-US', {
    year: 'numeric',
    month: 'short',
    day: 'numeric'
  });
}

/**
 * Get event type display name
 */
export function getEventTypeLabel(eventType: EventType): string {
  const labels: Record<EventType, string> = {
    CLINICAL_READOUT: 'Clinical Readout',
    MECHANISM_INSIGHT: 'Mechanism Insight',
    EVIDENCE_UPDATE: 'Evidence Update',
    REGULATORY_CHANGE: 'Regulatory Change',
    TRIAL_UPDATE: 'Trial Update',
    TARGET_VALIDATION: 'Target Validation',
    GENETIC_EVIDENCE: 'Genetic Evidence',
    BIOMARKER_DISCOVERY: 'Biomarker Discovery',
    PARTNERSHIP: 'Partnership',
    ACQUISITION: 'Acquisition',
    COMPETITIVE_INTEL: 'Competitive Intel',
    SAFETY_SIGNAL: 'Safety Signal',
    ENDPOINT_CHANGE: 'Endpoint Change',
    APPROVAL: 'Approval',
    REJECTION: 'Rejection',
    LABEL_UPDATE: 'Label Update'
  };

  return labels[eventType] || eventType;
}

/**
 * Get confidence level label
 */
export function getConfidenceLabel(score?: number): string {
  if (!score) return 'Unknown';
  if (score >= 0.8) return 'High';
  if (score >= 0.5) return 'Medium';
  return 'Low';
}

/**
 * Get impact level label
 */
export function getImpactLabel(score?: number): string {
  if (!score) return 'Unknown';
  if (score >= 0.7) return 'High';
  if (score >= 0.4) return 'Medium';
  return 'Low';
}

File: bt_platform/core/endpoints/science_events.py
"""
Science Event Store API Endpoints

Persistent, queryable, versioned science event/evidence store.
Provides canonical backend storage for:
- Clinical readouts
- Mechanism insights
- Evidence journals
- Regulatory updates
- Target validations

Supports historical querying, filtering, aggregation, versioning, and linkage.
"""

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from sqlalchemy import and_, or_, desc, func
from datetime import datetime, timedelta, date
from typing import Optional, List, Dict, Any
from ..database import get_db
from ..schema import ScienceEvent, Evidence, EventRelationship
from ..contracts import ScienceEventContract, EventRelationshipContract

router = APIRouter()


@router.post("/science-events", status_code=201)
async def create_science_event(
    event: ScienceEventContract,
    db: Session = Depends(get_db)
):
    """
    Create a new science event in the persistent store.

    Every event is stored as a discrete, versioned record with full provenance.
    """
    db_event = ScienceEvent(
        event_type=event.event_type,
        event_category=event.event_category,
        title=event.title,
        description=event.description,
        summary=event.summary,
        event_date=event.event_date,
        published_date=event.published_date,
        entity_type=event.entity_type,
        entity_id=event.entity_id,
        entity_name=event.entity_name,
        related_entities=event.related_entities,
        source_type=event.source_type,
        source_url=event.source_url,
        source_event_metadata=event.source_metadata,
        content=event.content,
        key_findings=event.key_findings,
        impact_assessment=event.impact_assessment,
        evidence_class=event.evidence_class,
        confidence_score=event.confidence_score,
        impact_score=event.impact_score,
        tags=event.tags,
        event_metadata=event.metadata,
        provider_file_sha256=event.provider_file_sha256,
        version=1,
        is_current=True
    )

    db.add(db_event)
    db.commit()
    db.refresh(db_event)

    return {
        "id": db_event.id,
        "event_type": db_event.event_type,
        "title": db_event.title,
        "event_date": db_event.event_date,
        "created_at": db_event.created_at
    }


@router.get("/science-events")
async def list_science_events(
    event_type: Optional[str] = Query(None, description="Filter by event type"),
    event_category: Optional[str] = Query(None, description="Filter by category"),
    entity_type: Optional[str] = Query(None, description="Filter by entity type"),
    entity_id: Optional[str] = Query(None, description="Filter by entity ID"),
    source_type: Optional[str] = Query(None, description="Filter by source type"),
    evidence_class: Optional[str] = Query(None, description="Filter by evidence class"),
    from_date: Optional[str] = Query(None, description="Start date (ISO format)"),
    to_date: Optional[str] = Query(None, description="End date (ISO format)"),
    tags: Optional[str] = Query(None, description="Comma-separated tags"),
    min_confidence: Optional[float] = Query(None, ge=0, le=1, description="Minimum confidence score"),
    min_impact: Optional[float] = Query(None, ge=0, le=1, description="Minimum impact score"),
    current_only: bool = Query(True, description="Only return current versions"),
    limit: int = Query(100, ge=1, le=1000),
    offset: int = Query(0, ge=0),
    db: Session = Depends(get_db)
):
    """
    Query science events with flexible filtering.

    Supports historical querying, filtering by date range, entity, source,
    confidence/impact scores, and tags.
    """
    query = db.query(ScienceEvent)

    # Apply filters
    if current_only:
        query = query.filter(ScienceEvent.is_current == True)

    if event_type:
        query = query.filter(ScienceEvent.event_type == event_type)

    if event_category:
        query = query.filter(ScienceEvent.event_category == event_category)

    if entity_type:
        query = query.filter(ScienceEvent.entity_type == entity_type)

    if entity_id:
        query = query.filter(ScienceEvent.entity_id == entity_id)

    if source_type:
        query = query.filter(ScienceEvent.source_type == source_type)

    if evidence_class:
        query = query.filter(ScienceEvent.evidence_class == evidence_class)

    if from_date:
        from_dt = datetime.fromisoformat(from_date.replace('Z', '+00:00'))
        query = query.filter(ScienceEvent.event_date >= from_dt)

    if to_date:
        to_dt = datetime.fromisoformat(to_date.replace('Z', '+00:00'))
        query = query.filter(ScienceEvent.event_date <= to_dt)

    if min_confidence is not None:
        query = query.filter(ScienceEvent.confidence_score >= min_confidence)

    if min_impact is not None:
        query = query.filter(ScienceEvent.impact_score >= min_impact)

    if tags:
        tag_list = [t.strip() for t in tags.split(',')]
        # Filter events that have any of the specified tags
        query = query.filter(
            or_(*[ScienceEvent.tags.contains([tag]) for tag in tag_list])
        )

    # Order by event date descending
    query = query.order_by(desc(ScienceEvent.event_date))

    # Get total count before pagination
    total = query.count()

    # Apply pagination
    events = query.offset(offset).limit(limit).all()

    return {
        "events": [
            {
                "id": event.id,
                "event_type": event.event_type,
                "event_category": event.event_category,
                "title": event.title,
                "summary": event.summary,
                "event_date": event.event_date,
                "published_date": event.published_date,
                "entity_type": event.entity_type,
                "entity_id": event.entity_id,
                "entity_name": event.entity_name,
                "source_type": event.source_type,
                "source_url": event.source_url,
                "evidence_class": event.evidence_class,
                "confidence_score": event.confidence_score,
                "impact_score": event.impact_score,
                "version": event.version,
                "is_current": event.is_current,
                "tags": event.tags,
                "created_at": event.created_at
            }
            for event in events
        ],
        "total": total,
        "limit": limit,
        "offset": offset
    }


@router.get("/science-events/{event_id}")
async def get_science_event(
    event_id: int,
    db: Session = Depends(get_db)
):
    """
    Get detailed information about a specific science event.
    """
    event = db.query(ScienceEvent).filter(ScienceEvent.id == event_id).first()

    if not event:
        raise HTTPException(status_code=404, detail="Science event not found")

    return {
        "id": event.id,
        "event_type": event.event_type,
        "event_category": event.event_category,
        "title": event.title,
        "description": event.description,
        "summary": event.summary,
        "event_date": event.event_date,
        "published_date": event.published_date,
        "entity_type": event.entity_type,
        "entity_id": event.entity_id,
        "entity_name": event.entity_name,
        "related_entities": event.related_entities,
        "source_type": event.source_type,
        "source_url": event.source_url,
        "source_metadata": event.source_metadata,
        "content": event.content,
        "key_findings": event.key_findings,
        "impact_assessment": event.impact_assessment,
        "evidence_class": event.evidence_class,
        "confidence_score": event.confidence_score,
        "impact_score": event.impact_score,
        "version": event.version,
        "parent_version_id": event.parent_version_id,
        "is_current": event.is_current,
        "change_summary": event.change_summary,
        "tags": event.tags,
        "metadata": event.metadata,
        "created_at": event.created_at,
        "updated_at": event.updated_at
    }


@router.get("/science-events/{event_id}/history")
async def get_event_history(
    event_id: int,
    db: Session = Depends(get_db)
):
    """
    Get the version history of a science event.

    Returns all versions of the event in chronological order.
    """
    # Get the current event
    current_event = db.query(ScienceEvent).filter(
        ScienceEvent.id == event_id
    ).first()

    if not current_event:
        raise HTTPException(status_code=404, detail="Science event not found")

    # If this is a version, find the root
    root_id = event_id
    while current_event.parent_version_id is not None:
        root_id = current_event.parent_version_id
        current_event = db.query(ScienceEvent).filter(
            ScienceEvent.id == root_id
        ).first()

    # Get all versions starting from root
    versions = []
    queue = [root_id]
    visited = set()

    while queue:
        current_id = queue.pop(0)
        if current_id in visited:
            continue
        visited.add(current_id)

        event = db.query(ScienceEvent).filter(ScienceEvent.id == current_id).first()
        if event:
            versions.append(event)
            # Find child versions
            children = db.query(ScienceEvent).filter(
                ScienceEvent.parent_version_id == current_id
            ).all()
            queue.extend([child.id for child in children])

    # Sort by version number
    versions.sort(key=lambda x: x.version)

    return {
        "versions": [
            {
                "id": v.id,
                "version": v.version,
                "title": v.title,
                "event_date": v.event_date,
                "is_current": v.is_current,
                "change_summary": v.change_summary,
                "created_at": v.created_at
            }
            for v in versions
        ]
    }


@router.put("/science-events/{event_id}")
async def update_science_event(
    event_id: int,
    event: ScienceEventContract,
    change_summary: Optional[str] = Query(None, description="Summary of changes"),
    db: Session = Depends(get_db)
):
    """
    Update a science event by creating a new version.

    Preserves the old version for historical tracking.
    """
    # Get the current event
    current_event = db.query(ScienceEvent).filter(
        ScienceEvent.id == event_id,
        ScienceEvent.is_current == True
    ).first()

    if not current_event:
        raise HTTPException(status_code=404, detail="Science event not found or not current")

    # Mark current event as not current
    current_event.is_current = False

    # Create new version
    new_event = ScienceEvent(
        event_type=event.event_type,
        event_category=event.event_category,
        title=event.title,
        description=event.description,
        summary=event.summary,
        event_date=event.event_date,
        published_date=event.published_date,
        entity_type=event.entity_type,
        entity_id=event.entity_id,
        entity_name=event.entity_name,
        related_entities=event.related_entities,
        source_type=event.source_type,
        source_url=event.source_url,
        source_event_metadata=event.source_metadata,
        content=event.content,
        key_findings=event.key_findings,
        impact_assessment=event.impact_assessment,
        evidence_class=event.evidence_class,
        confidence_score=event.confidence_score,
        impact_score=event.impact_score,
        tags=event.tags,
        event_metadata=event.metadata,
        provider_file_sha256=event.provider_file_sha256,
        version=current_event.version + 1,
        parent_version_id=event_id,
        is_current=True,
        change_summary=change_summary
    )

    db.add(new_event)
    db.commit()
    db.refresh(new_event)

    return {
        "id": new_event.id,
        "version": new_event.version,
        "parent_version_id": new_event.parent_version_id,
        "created_at": new_event.created_at
    }


@router.get("/science-events/timeline/{entity_type}/{entity_id}")
async def get_entity_timeline(
    entity_type: str,
    entity_id: str,
    from_date: Optional[str] = Query(None, description="Start date (ISO format)"),
    to_date: Optional[str] = Query(None, description="End date (ISO format)"),
    event_types: Optional[str] = Query(None, description="Comma-separated event types"),
    db: Session = Depends(get_db)
):
    """
    Get a timeline of science events for a specific entity.

    Perfect for building evidence timelines and science dashboards.
    """
    query = db.query(ScienceEvent).filter(
        ScienceEvent.entity_type == entity_type,
        ScienceEvent.entity_id == entity_id,
        ScienceEvent.is_current == True
    )

    if from_date:
        from_dt = datetime.fromisoformat(from_date.replace('Z', '+00:00'))
        query = query.filter(ScienceEvent.event_date >= from_dt)

    if to_date:
        to_dt = datetime.fromisoformat(to_date.replace('Z', '+00:00'))
        query = query.filter(ScienceEvent.event_date <= to_dt)

    if event_types:
        type_list = [t.strip() for t in event_types.split(',')]
        query = query.filter(ScienceEvent.event_type.in_(type_list))

    events = query.order_by(ScienceEvent.event_date).all()

    return {
        "entity_type": entity_type,
        "entity_id": entity_id,
        "timeline": [
            {
                "id": event.id,
                "event_type": event.event_type,
                "event_category": event.event_category,
                "title": event.title,
                "summary": event.summary,
                "event_date": event.event_date,
                "source_type": event.source_type,
                "source_url": event.source_url,
                "evidence_class": event.evidence_class,
                "confidence_score": event.confidence_score,
                "impact_score": event.impact_score,
                "key_findings": event.key_findings
            }
            for event in events
        ],
        "total_events": len(events)
    }


@router.post("/event-relationships", status_code=201)
async def create_event_relationship(
    relationship: EventRelationshipContract,
    db: Session = Depends(get_db)
):
    """
    Create a relationship between two science events.

    Useful for building knowledge graphs and understanding event connections.
    """
    # Verify both events exist
    source = db.query(ScienceEvent).filter(ScienceEvent.id == relationship.source_event_id).first()
    target = db.query(ScienceEvent).filter(ScienceEvent.id == relationship.target_event_id).first()

    if not source or not target:
        raise HTTPException(status_code=404, detail="One or both events not found")

    db_rel = EventRelationship(
        source_event_id=relationship.source_event_id,
        target_event_id=relationship.target_event_id,
        relationship_type=relationship.relationship_type,
        description=relationship.description,
        confidence=relationship.confidence,
        event_metadata=relationship.metadata
    )

    db.add(db_rel)
    db.commit()
    db.refresh(db_rel)

    return {
        "id": db_rel.id,
        "source_event_id": db_rel.source_event_id,
        "target_event_id": db_rel.target_event_id,
        "relationship_type": db_rel.relationship_type,
        "created_at": db_rel.created_at
    }


@router.get("/event-relationships/{event_id}")
async def get_event_relationships(
    event_id: int,
    relationship_type: Optional[str] = Query(None, description="Filter by relationship type"),
    direction: str = Query("both", regex="^(incoming|outgoing|both)$"),
    db: Session = Depends(get_db)
):
    """
    Get relationships for a science event.

    Returns incoming, outgoing, or both relationship directions.
    """
    query_conditions = []

    if direction in ["incoming", "both"]:
        query_conditions.append(EventRelationship.target_event_id == event_id)

    if direction in ["outgoing", "both"]:
        query_conditions.append(EventRelationship.source_event_id == event_id)

    query = db.query(EventRelationship).filter(or_(*query_conditions))

    if relationship_type:
        query = query.filter(EventRelationship.relationship_type == relationship_type)

    relationships = query.all()

    return {
        "event_id": event_id,
        "relationships": [
            {
                "id": rel.id,
                "source_event_id": rel.source_event_id,
                "target_event_id": rel.target_event_id,
                "relationship_type": rel.relationship_type,
                "description": rel.description,
                "confidence": rel.confidence,
                "created_at": rel.created_at
            }
            for rel in relationships
        ]
    }


@router.get("/science-events/aggregate/by-type")
async def aggregate_events_by_type(
    from_date: Optional[str] = Query(None, description="Start date (ISO format)"),
    to_date: Optional[str] = Query(None, description="End date (ISO format)"),
    entity_type: Optional[str] = Query(None, description="Filter by entity type"),
    db: Session = Depends(get_db)
):
    """
    Aggregate science events by type for dashboard views.
    """
    query = db.query(
        ScienceEvent.event_type,
        func.count(ScienceEvent.id).label('count')
    ).filter(ScienceEvent.is_current == True)

    if from_date:
        from_dt = datetime.fromisoformat(from_date.replace('Z', '+00:00'))
        query = query.filter(ScienceEvent.event_date >= from_dt)

    if to_date:
        to_dt = datetime.fromisoformat(to_date.replace('Z', '+00:00'))
        query = query.filter(ScienceEvent.event_date <= to_dt)

    if entity_type:
        query = query.filter(ScienceEvent.entity_type == entity_type)

    results = query.group_by(ScienceEvent.event_type).all()

    return {
        "aggregations": [
            {
                "event_type": result.event_type,
                "count": result.count
            }
            for result in results
        ]
    }


@router.get("/science-events/search")
async def search_science_events(
    q: str = Query(..., min_length=3, description="Search query"),
    limit: int = Query(50, ge=1, le=200),
    db: Session = Depends(get_db)
):
    """
    Full-text search across science events.

    Searches title, description, summary, and content fields.
    """
    search_pattern = f"%{q}%"

    events = db.query(ScienceEvent).filter(
        ScienceEvent.is_current == True,
        or_(
            ScienceEvent.title.ilike(search_pattern),
            ScienceEvent.description.ilike(search_pattern),
            ScienceEvent.summary.ilike(search_pattern),
            ScienceEvent.content.ilike(search_pattern)
        )
    ).order_by(desc(ScienceEvent.event_date)).limit(limit).all()

    return {
        "query": q,
        "results": [
            {
                "id": event.id,
                "event_type": event.event_type,
                "title": event.title,
                "summary": event.summary,
                "event_date": event.event_date,
                "entity_type": event.entity_type,
                "entity_id": event.entity_id,
                "source_type": event.source_type
            }
            for event in events
        ],
        "count": len(events)
    }
